context_slog.go

```diff
@@ -0,0 +1,83 @@
+//go:build go1.21
+// +build go1.21
+
+/*
+Copyright 2019 The logr Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package logr
+
+import (
+	"context"
+	"fmt"
+	"log/slog"
+)
+
+// FromContext returns a Logger from ctx or an error if no Logger is found.
+func FromContext(ctx context.Context) (Logger, error) {
+	v := ctx.Value(contextKey{})
+	if v == nil {
+		return Logger{}, notFoundError{}
+	}
+
+	switch v := v.(type) {
+	case Logger:
+		return v, nil
+	case *slog.Logger:
+		return NewLogr(v.Handler()), nil
+	default:
+		// Not reached.
+		panic(fmt.Sprintf("unexpected value type for logr context key: %T", v))
+	}
+}
+
+// FromContext returns a slog.Logger from ctx or nil if no such Logger is found.
+func SlogLoggerFromContext(ctx context.Context) *slog.Logger {
```

Comment:
I decided to focus on handling a `slog.Logger` pointer instead of a `slog.Handler` because a) that is what [slog-context](https://github.com/veqryn/slog-context) does and b) because only supporting `slog.Handler` would imply one `slog.New` on each retrieval.

It would also be possible to extend this to `slog.Handler`.

Reply:
In a way, the whole point of context support for slog is to enable a familiar API for developers - that is `slog.Logger`, not `slog.Handler`.

I still think developers should be more flexible and just use `logr.Logger`... :shrug:

Reply:
👍